src/apps/vle/mvle.1
.TH "MVLE" "1"

.SH "NAME"
mvle \(em The MPI experimental frames executor for VLE framework

.SH "SYNOPSIS"
.PP
\fBmvle\fR
[\fB-h\fP, \fB\-\-help\fP]
[\fB\-P\fP, \fB\-\-package \fIpackage_name\fP\fR]
[\fB\-v\fP]
[\fB\-\-version\fP]
\fB\fIvpz files\fP...

.SH "DESCRIPTION"
.PP
This manual page documents briefly the \fBmvle\fR command.
.PP
\fBMVLE\fR is a part of a multi-modelling and simulation framework call
\fBVLE\fR. It addresses the reliability issue using recent developments in the
theory of Modelling and Simulation proposed by Zeigler and open sources
software. \fBVLE\fR is composed of five applications: \fIVLE\fR the simulator,
\fIGVLE\fR the modelling tool.
.PP
\fBMVLE\fR is a program to execute experimental frame on classical cluster
system: \fBMPI\fR (Message Parsing Interface).

.SH "OPTIONS"
.PP
These programs follow the usual GNU command line syntax, with long options
starting with two dashes (`\-'). A summary of options is included below.

.IP "\fB-h\fP, \fB\-\-help\fP" 10
Show summary of options.

.IP "\fB-v\fp, \fB\-\-version\fP" 10
Show version of program.

.IP "\fB-P\fP, \fB\-\-package\fI packagename\fR\fP"
Selects the VLE package where search experimental frame from the $VLE_HOME
directory.

.SH "EXAMPLES"
.PP
Run mvle on 32 process, for the experimental frame `firemanqss-exp.vpz' of the
package `firemanqss':
.PP
$ mpirun -np 32 mvle -P firemanqss firemanqss-exp.vpz

.PP
Run mvle on 2048 processor from a specified machine file, for the experimental
frame `firemanqss-exp.vpz' of the package `firemanqss':
.PP
$ mpirun -np 2048 --machinefile file.txt mvle -P vle.examples unittest.vpz

.SH "ENVIRONMENTS"
.IP VLE_HOME
A path where you push models packages (ie. simulators, streams and modelling
plugins,  vpz files, data, and the outputs of simulaton)

.SH "SEE ALSO"
.PP
vle (1), oov (1) and gvle (1).

.SH "AUTHOR"
.PP
Gauthier Quesnel \fBquesnel@users.sourceforge.net\fP
.PP
This program is free software: you can redistribute it and/or modify it under
the terms of the GNU General Public License as published by the Free Software
Foundation, either version 3 of the License, or (at your option) any later
version.

.SH "COPYRIGHT"
.PP
Copyright (c) 2003-2007 Gauthier Quesnel <quesnel@users.sourceforge.net>
.PP
Copyright (c) 2003-2011 ULCO http://www.univ-littoral.fr
.PP
Copyright (c) 2007-2011 INRA http://www.inra.fr


NEWS
VLE 1.1.0-dev
=============

We merge all shared libraries of vle (libvleutils, libvlevpz,
libvledevs etc.) into a uniq shared library libvle.so.

We remove the Eov program and the libvleeov shared library. Now,
graphical output are available into the GVLE application.

- libvleutils:

  - Add a new RemoteMananger class to access remote repositories. The
    command line interface is inspirited from the apt-get debian's
    command:

  $ vle --remote update
  $ vle --remote install glue
  $ vle --remote search '*gl*'
  $ vle --remote show glue

- libvlegeometry: removed

- libvleeov: removed

VLE 1.0.1
=========

A debug release

Application
~~~~~~~~~~~

vle: switch to copyright to 2011


VFL (VLE Foundation Libraries)
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

- extension:
  - decision:
    - Fix behavior of the SS constraint: A constraint SS(A,B,mintl,maxtl) could
      lead to a start of the activity B when A is done. B should start only
      mintl after A is done.
    - Add parametrized error messages: To produce more explicit error messages,
      the vle::fmt method is used to parametrize the modelling error messages.
  - de:
    - Fix a bug on confluent transition: When a Nosync variable arrives on
      external transition, a useless transition to POST state with ta=0 is set,
      even if all synchronous variables are available. The test example
      confluentNosync is added. (Closes: #3409820).
  - dess:
    - Fix the example model sir4.vpz: Change of output name of the view (Closes:
      #3425845).
- utils:
  - Fix the paths of external plug-ins: The Path API for external plug-ins of
    GVLE output and modelling directories are wrongs. We fix it.
    - Path::getExternalPackagePluginGvleOutputDir returns
      $package/plugins/gvle/modeling
    - Path::getExternalPackagePluginGvleModelingDir returns
      $package/plugins/gvle/output


VLE 1.0.0
=========

A final release

Application
~~~~~~~~~~~

- gvle:
  - add the item "Select All" in to the "Edit" menu.
  - use Modeling like the M of MVC.
  - add a tooltip for the experimental conditions.
  - select the file of the focused tab & vice versa.
  - update the OpenPackageBox Dialog.
  - fix segfault in FileTreeView::getFileRow.
  - fix segfault in Editor when used by plugins.


VLE 1.0.0-alpha6
================

A debug release

Application
~~~~~~~~~~~

- gvle:
  - Add multiple selection to the Observable Editor:
    - Observables and Views columns are now multiple selectable in
      ObsAndViewBox.
    - Treatments have beed added to manage many rows selected in each columns.
      (Closes: #3290348)
  - Update the view selection mechanism:
    - Remove the Drag'n Drop system of the ObsAndViewBox dialog.
    - Add 2 buttons to add and delete Views in Observables. (Closes: #3290338)
  - Add an include field in DE plug-in:
    - Add an Include text field in DifferenceEquation.glade.
    - Add a variable "mInclude" in "SourceDialog" class and linked treatments.
    - Edit Generic, Multiple and Simple classes to manage "mInclude" in order
      to insert it in the ".cpp" generated file. (Closes: #3285682)
  - Fix editing experimental condition:
    - The ConditionBox class manage and return a list of renamed experimental
      condition.
    - In AtomicModelBox and GVLE classes a treatment to manage the
      ConditionBox's renameList has been added.
    - When a experimental condition is renamed, a treatment to update all of
      the project models and class models is provided.
    - Add a method applyRemoved in AtomicModelBox and GVLE classes to manage
      the removing condition case in all of project models and class models.
      (Closes: #3009808)


VLE 1.0.0-alpha5
================

A release to fix bug with a new Win32 environment variables system.


Applications
~~~~~~~~~~~~

- gvle:
  - Fix default size on value box. When the user create a new set or map value,
    the dialog box is too small and the user must resize the dialog box if its
    wants to add a new value (Closes: #3289724).
  - Add a julian-day parameter on file output plugin (Closes: #3279408).
  - Sort the library list in DynamicBox (Closes: #3010376).
  - Fix source port in link tooltip information (Closes: #3279204).
  - Fix the modeling plug-in launching.
  - Fix the logo loading on win32.
  - Fix segfault when adding new view (Closes: #3257511, #3240925, #3202834).

- win32:
  - Replace the win32 environment variable system: We remove the Win32
    environment variable from the NSIS script. We add these variables
    dynamically, before calling the Package methods' configure(), build(),
    clean() etc.


VFL (VLE Foundation Libraries)
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

- libdevs: fix crash when throwing exception in model's: Since the update of
  the value::Map class, the experimental conditions are not cloned but just
  referenced. Thus, we must to clear the experimental condition data instead of
  deleting the data causing a double free memory. We fix the problem by
  clearing the value::Map instead of deleting it (Closes: #3288773, #3184804).

- libvleextension:
  - ext/qss: fix bug on interaction with de
    - If a variable of difference equation is in interaction with QSS model and
      the QSS model have several continuous variables, the QSS model fails
      because its waits the discrete variable at init state.
    - Add a control on input port name: the possible names are 'update' and
      'perturb'.
  - ext/de:
    - Fix test with tolerance

- libvleutils:
  - Fix missing header in utils.hpp
  - Switch Trace singleton in thread-safe way: The utils::Trace singleon is not
    thread safe. We transform the utils::Trace class into a static class which
    manage a Pimpl class to ensure the thead-safe.

- libvlevpz:
  - Fix sax parser when dealing with Table The parsing of a value::Table was
    not performed correctly and can throw an error. We fix the parsing of table
    xml node element and we improve the runtime checking of table width and

src/apps/vle/mvle.cpp
/*
 * @file apps/vle/mvle.cpp
 *
 * This file is part of VLE, a framework for multi-modeling, simulation
 * and analysis of complex dynamical systems
 * http://www.vle-project.org
 *
 * Copyright (c) 2003-2007 Gauthier Quesnel <quesnel@users.sourceforge.net>
 * Copyright (c) 2003-2011 ULCO http://www.univ-littoral.fr
 * Copyright (c) 2007-2011 INRA http://www.inra.fr
 *
 * See the AUTHORS or Authors.txt file for copyright owners and contributors
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */


#include <vle/manager/RunManager.hpp>
#include <vle/manager/Manager.hpp>
#include <vle/utils/Tools.hpp>
#include <vle/utils/Path.hpp>
#include <vle/utils/Package.hpp>
#include <vle/utils/i18n.hpp>
#include <vle/version.hpp>
#include <cstdio>
#include <cstdarg>
#include <cstring>

#define OMPI_SKIP_MPICXX
#include <mpi.h>

void mvle_show_help()
{
    std::fprintf(stderr, _(
            "Use:\n"
            "  mvle [-h,--help] [-v,--version] [-s|--show] [-P,--package"
            " package_name] vpz_files...\n"
            "\n"
            "Help options:\n"
            "  -h, --help        Show help option\n"
            "\n"
            "Application options:\n"
            "  -s --show         Show the plan\n"
            "  -P --package      Start VLE in package mode\n"
            "  -v --version      Show the version\n"));
}

void mvle_show_version()
{
    std::fprintf(stderr, _(
            "mvle - MPI manager for VLE\n"
            "Virtual Laboratory Environment - %s\n"
            "Copyright (C) 2003 - 2011 The VLE Development Team.\n"
            "VLE is a multi-modeling environment to build,\nsimulate "
            "and analyse models of dynamic complex systems.\n"
            "For more information, see manuals with 'man mvle' or\n"
            "the VLE website http://sourceforge.net/projects/vle/\n"),
        VLE_NAME_COMPLETE);
}

void mvle_print(const char *fmt, ...)
{
    va_list argptr;

    va_start(argptr, fmt);
    std::vfprintf(stdout, fmt, argptr);
    va_end(argptr);
}

void mvle_print_debug(const char *fmt, ...)
{
    std::fprintf(stderr, "mvle_debug: ");

    va_list argptr;

    va_start(argptr, fmt);
    std::vfprintf(stderr, fmt, argptr);
    va_end(argptr);

    std::fprintf(stderr, "\n");
}

void mvle_print_error(const char *fmt, ...)
{
    std::fprintf(stderr, "mvle_error: ");

    va_list argptr;

    va_start(argptr, fmt);
    std::vfprintf(stderr, fmt, argptr);
    va_end(argptr);

    std::fprintf(stderr, "\n");
}

void mvle_mpi_error(int error)
{
    int errorclass;
    int len;
    char *buffer = new char[MPI_MAX_ERROR_STRING + 1];
    buffer[MPI_MAX_ERROR_STRING] = '\0';

    MPI_Error_class(error, &errorclass);
    MPI_Error_string(errorclass,  buffer, &len);
    buffer[len] = '\0';

    mvle_print_error("%s", buffer);
}

bool mvle_mpi_init(int *argc, char ***argv, uint32_t *rank, uint32_t *world)
{
    int t_rank;
    int t_world;
    int r;
    bool result = false;

    if ((r = MPI_Init(argc, argv)) == MPI_SUCCESS) {
        if ((r = MPI_Comm_rank(MPI_COMM_WORLD, &t_rank)) == MPI_SUCCESS) {
            /* check the cast of the MPI's rank */
            if (t_rank < 0 or static_cast < unsigned int >(t_rank) >
                std::numeric_limits < uint32_t >::max()) {
                mvle_print_error(_("bad mpi rank id: %d"), t_rank);
            } else {
                *rank = static_cast < uint32_t >(t_rank);
                if ((r = MPI_Comm_size(MPI_COMM_WORLD, &t_world)) ==
                    MPI_SUCCESS) {
                    /* check the cast of the MPI's world size */
                    if (t_world < 0 and static_cast < unsigned int >(t_world) >
                        std::numeric_limits < uint32_t >::max()) {
                        mvle_print_error(_("bad mpi world size: %d"), t_world);
                    } else {
                        *world = static_cast < uint32_t >(t_world);
                        result = true;
                    }
                }
            }
        }
    }

    if (r != MPI_SUCCESS) {
        mvle_mpi_error(r);
        result = false;
    }

    return result;
}

bool mvle_parse_arg(int argc, char **argv, int *vpz, bool *show)
{
    int i = 1;

    while (i < argc) {
        if ((std::strcmp(argv[i], "-P") == 0 or
             std::strcmp(argv[i], "--package") == 0) and i + 1 <= argc) {
            vle::utils::Package::package().select(argv[++i]);
        } else if (std::strcmp(argv[i], "-h") == 0 or
                   std::strcmp(argv[i], "--help") == 0) {
            mvle_show_help();
            return false;
        } else if (std::strcmp(argv[i], "-v") == 0 or
                   std::strcmp(argv[i], "--version") == 0) {
            mvle_show_version();
            return false;
        } else if (std::strcmp(argv[i], "-s") == 0 or
                   std::strcmp(argv[i], "--show") == 0) {
            *show = true;
        } else {
            *vpz = i;
        }
        ++i;
    }

    return *vpz < argc and not vle::utils::Package::package().name().empty();
}

void mvle_show(const std::string& vpz)
{
    vle::manager::ExperimentGenerator expgen(vpz, 0, 1);

    if (expgen.size()) {
        vle::vpz::Conditions conds;
        vle::vpz::Conditions::iterator it;

        expgen.get(expgen.min(), &conds);

        mvle_print(";");
        for (it = conds.begin(); it != conds.end(); ++it) {
            const vle::vpz::Condition& cond = it->second;
            vle::vpz::Condition::const_iterator jt;

            for (jt = cond.begin(); jt != cond.end(); ++jt) {
                mvle_print("%s.%s;", it->first.c_str(), jt->first.c_str());
            }
        }

        mvle_print("\n");

        for (uint32_t i = expgen.min(); i < expgen.max(); ++i) {
            expgen.get(i, &conds);

            mvle_print("%d;", i);
            for (it = conds.begin(); it != conds.end(); ++it) {
                const vle::vpz::Condition& cond = it->second;
                vle::vpz::Condition::const_iterator jt;

                for (jt = cond.begin(); jt != cond.end(); ++jt) {
                    mvle_print("%s;",
                               jt->second->get(0)->writeToFile().c_str());
                }
            }
            mvle_print("\n");
        }
    }
}

int main(int argc, char **argv)
{
    uint32_t rank = 0;
    uint32_t world = 0;
    bool show = false;
    bool result = true;

    vle::manager::init();

    if ((result = mvle_mpi_init(&argc, &argv, &rank, &world))) {
        int vpz;

        if ((result = mvle_parse_arg(argc, argv, &vpz, &show))) {
            if (show) {
                while (vpz < argc) {
                    mvle_show(
                        vle::utils::Path::path().
                        getPackageExpFile(argv[vpz]));
                    vpz++;
                }
            } else {
                try {
                    vle::manager::RunManager man(rank, world);
                    mvle_print("MPI node %d/%d start\n", rank, world);

                    while (vpz < argc) {
                        result = man.run(
                            vle::utils::Path::path().
                            getPackageExpFile(argv[vpz]));
                        vpz++;
                    }
                } catch (const std::exception& e) {
                    mvle_print_error("manager problem: %s", e.what());
                    result = false;
                }
            }
        }
    }

    vle::manager::finalize();

    MPI_Finalize();

    return result ? EXIT_SUCCESS : EXIT_FAILURE;
}
